### ISSUE TITLE
How to change lj14scale in NonbondedForce?

### ISSUE THREAD

**Issue by asgharrazavi** [openforcefield/openff-toolkit#1516]
I am trying to use openff's parameters in my simulations using OpenMM. However, I am getting error that my ligand parameters (coming from openff-toolkit) have a different lj14scale scaling than openMM's native forcefield files. In short here is the lj14scale from openff  `<NonbondedForce coulomb14scale="1.0" lj14scale="1.0">` and here is the lj14scale from openMM `  <NonbondedForce coulomb14scale="0.8333333333333334" lj14scale="0.5">`. 
Is there a way to convert lj14scales coming from openff?

**Comment by mattwthompson** [comment]
There are a few different ways this could happen, some reconcilable and some not. Could you share a reproduction?

**Comment by asgharrazavi** [comment]
I am using this code to get my ligand xml file:
```import mdtraj as md
from openff.interchange import Interchange
from openmm import app
import parmed

from openff.toolkit import ForceField, Molecule
from openff.toolkit.utils import get_data_file_path

lig = Molecule.from_file('6126_lig.sdf')
sage = ForceField("openff-2.0.0.offxml")
f_lig = Interchange.from_smirnoff(force_field=sage, topology=lig.to_topology())

st = parmed.openmm.load_topology(lig.to_topology().to_openmm(), system=f_lig.to_openmm())
w =  parmed.amber.parameters.ParameterSet.from_structure(st)
ww = parmed.openmm.parameters.OpenMMParameterSet.from_parameterset(w)
ww.residues.update(parmed.modeller.ResidueTemplateContainer.from_structure(st).to_library())
ww.write('ligand.xml')
```

When I try to use it in openMM using the code below, I get the incompatibility error:

```
from openmm import app
omm_forcefield = app.ForceField('amber14-all.xml', 'amber14/tip3pfb.xml','./ligand.xml')
```

Error message:
```Traceback (most recent call last):
    omm_forcefield = app.ForceField('amber14-all.xml', 'amber14/tip3pfb.xml','./lig.xml')
  File "/Users/asgharrazavi/miniconda3/envs/openff/lib/python3.9/site-packages/openmm/app/forcefield.py", line 214, in __init__
    self.loadFile(files)
  File "/Users/asgharrazavi/miniconda3/envs/openff/lib/python3.9/site-packages/openmm/app/forcefield.py", line 407, in loadFile
    parsers[child.tag](child, self)
  File "/Users/asgharrazavi/miniconda3/envs/openff/lib/python3.9/site-packages/openmm/app/forcefield.py", line 2446, in parseElement
    raise ValueError('Found multiple NonbondedForce tags with different 1-4 scales')
ValueError: Found multiple NonbondedForce tags with different 1-4 scales
```

**Comment by asgharrazavi** [comment]
Thanks for the code and explanations. I will post the question in parmed channel. 
In the meantime do you know of an approach to save parameters generated by openff in the `xml` format without using parmed?

**Comment by mattwthompson** [comment]
You can serialize the `openmm.System` object directly with OpenMM:

```python3
In [1]: from openff.toolkit import ForceField, Molecule

In [2]: from openmm import XmlSerializer

In [3]: openmm_system = ForceField("openff-2.0.0.offxml").create_openmm_system(
   ...:     Molecule.from_smiles("CCO").to_topology()
   ...: )

In [4]: with open("system.xml", "w") as f:
   ...:     f.write(XmlSerializer.serialize(openmm_system))

In [5]: !head system.xml
<?xml version="1.0" ?>
<System openmmVersion="7.7" type="System" version="1">
	<PeriodicBoxVectors>
		<A x="2" y="0" z="0"/>
		<B x="0" y="2" z="0"/>
		<C x="0" y="0" z="2"/>
	</PeriodicBoxVectors>
	<Particles>
		<Particle mass="12.01078"/>
		<Particle mass="12.01078"/>
```

http://docs.openmm.org/latest/api-python/generated/openmm.openmm.XmlSerializer.html?highlight=xmlserializer#openmm.openmm.XmlSerializer.serialize

**Comment by asgharrazavi** [comment]
I tried the `XmlSerializer` but the problem is that it doesn't have residue and atom information e.g. residue name or atom names and types, hence, openMM can't understand how to use this xml file to extract forcefield parameters for a molecule.
This is the error form OpenMM:

```ValueError: No template found for residue 295 (ENK).  This might mean your input topology is missing some atoms or bonds, or possibly that you are using the wrong force field.```

**Comment by asgharrazavi** [comment]
Thanks! I used approach 2 that you suggested in the code below and it seems that simulations are running fine. 

```python
from openmm import *
from openmm import unit
from openmm import app
import numpy as np
import os,sys
import time
from openff.toolkit import Molecule
from openmmforcefields.generators import SMIRNOFFTemplateGenerator


print('building protein and solvating...')
omm_forcefield = app.ForceField('amber14-all.xml', 'amber14/tip3pfb.xml')
prot_pdb_file = app.PDBFile('6126_prot_lig.pdb') 
top = prot_pdb_file.topology
    
lig = Molecule.from_file('6126_lig.sdf')
smirnoff = SMIRNOFFTemplateGenerator(forcefield="openff-2.0.0", molecules=lig)
omm_forcefield.registerTemplateGenerator(smirnoff.generator)

modeller = app.Modeller(top, prot_pdb_file.positions)
    
print('Adding hydrogens...')
modeller.addHydrogens(omm_forcefield)
    
print('Adding solvent...')
modeller.addSolvent(omm_forcefield, model='tip3p', padding=0.9*unit.nanometer, ionicStrength=0.1*unit.molar, positiveIon='Na+', negativeIon='Cl-')
    
modeller.topology.setUnitCellDimensions(modeller.topology.getUnitCellDimensions())
        
positions = modeller.positions
topology = modeller.topology
    
print('Building system...')
system = omm_forcefield.createSystem(topology, nonbondedMethod=app.PME, nonbondedCutoff=0.9*unit.nanometer, constraints=app.HBonds) 
    
barostat = MonteCarloBarostat(1.0*unit.bar, 310.0*unit.kelvin)
system.addForce(barostat)
    
integrator = LangevinMiddleIntegrator(310*unit.kelvin, 1/unit.picosecond, 0.002*unit.picoseconds)
simulation = app.Simulation(topology, system, integrator)
simulation.minimizeEnergy()
```

I notices that the building system is not taking a long time, so I am worries whether the small molecule parametrization is being done properly. Is there a way to compare the ligand parameters from code above to the parameters I am getting from code below?

```python
from openff.interchange import Interchange
from openmm import app
import parmed

from openff.toolkit import ForceField, Molecule
from openff.toolkit.utils import get_data_file_path

lig = Molecule.from_file('6126_lig.sdf')
sage = ForceField("openff-2.0.0.offxml")
f_lig = Interchange.from_smirnoff(force_field=sage, topology=lig.to_topology())
```

**Comment by j-wags** [comment]
Are you sure your ligand is in the complex file (`6126_prot_lig.pdb`)?

**Comment by j-wags** [comment]
For context - I'm working on reproducing this locally and verifying that the same ligand parameters appear in the final OpenMM system in both cases. And it took me the longest time to notice that the PDB I was loading lacked the ligand. 

So if you could send your inputs that would be very helpful!

**Comment by asgharrazavi** [comment]
yes, it is. I ran short simulations and load the trajectory and I can see the ligand. The ligand is moving a bit more than I expected.

**Comment by asgharrazavi** [comment]
I won't be able to send the exact system that I am using, but will try and find a general system as an example and send you the files.

**Comment by j-wags** [comment]
Thanks. Using a random system from my local files has modeller raising an error, so I realized I was getting bogged down in things that may not be relevant to your issue. So getting something that runs like your inputs will be really helpful.

And to skip ahead - if there's a big difference in how long system creation is taking, that makes me think that something is wrong with the AM1BCC charge assignment (usually the slow part of system creation). So to check whether the parameterization really is equivalent I'd recommend pulling some random charges from the `f_lig` Interchange object and seeing if they're also present somewhere in the `system` object that comes out of the other pathway. If I could reproduce the issue I'd use code like this to check:

```python
# get the partial charge from some random ligand atom from pathway 2
print(list(f_lig.handlers['Electrostatics'].get_charges().values())[1])
```
>-0.6031100149266422 elementary_charge

```python
# Make sure that there's some particle in the combined OpenMM system that has a charge like those from other workflow
from openmm import XmlSerializer
xml = XmlSerializer.serialize(system)
# Note that OpenMM systems may round/format ligand charges differently, so I modify the charge from above a bit to lose the leading 0 and decrease sig figs
idx = xml.find('q="-.6031')
print(idx)
print(xml[idx-10:idx+20])
```
> hopefully some number other than -1
> hopefully some 30-character snippet of the OpenMM system XML that has the nonbonded parameters for the corresponding atom

**Comment by asgharrazavi** [comment]
Here is an example of a system and script that I am using to parametrize and run the simulations. If you notice any issues in forcefield parametrization please let me know. Thanks. 

ligand file: https://github.com/asgharrazavi/adaptive_sampling/blob/master/test_vhl_6zhc_lig.sdf
protein+ligand file: https://github.com/asgharrazavi/adaptive_sampling/blob/master/test_vhl_6zhc_prot_lig.pdb
run script: https://github.com/asgharrazavi/adaptive_sampling/blob/master/run_openmm.py

**Comment by asgharrazavi** [comment]
Looks like the script is working fine when I have only one ligand, however, when I have two ligands, it fails parametrization for the second ligand. Here is the script: 
```python 
lig1 = Molecule.from_file('lig1.sdf')
lig2 = Molecule.from_file('lig2.sdf')
smirnoff1 = SMIRNOFFTemplateGenerator(forcefield="openff-2.0.0", molecules=lig1)
smirnoff2 = SMIRNOFFTemplateGenerator(forcefield="openff-2.0.0", molecules=lig2)
omm_forcefield = app.ForceField('amber14-all.xml', 'amber14/tip3pfb.xml')
print('adding first ligand')
omm_forcefield.registerTemplateGenerator(smirnoff1.generator)
print('adding second ligand')
omm_forcefield.registerTemplateGenerator(smirnoff2.generator)
```
your procedure below:

```python
# Make sure that there's some particle in the combined OpenMM system that has a charge like those from other workflow
from openmm import XmlSerializer
xml = XmlSerializer.serialize(system)
# Note that OpenMM systems may round/format ligand charges differently, so I modify the charge from above a bit to lose the leading 0 and decrease sig figs
idx = xml.find('q="-.6031')
print(idx)
print(xml[idx-10:idx+20])
```
is returning `-1` for the second ligand. 

Also openMM issues this warning: `Did not recognize residue PLP; did you forget to call .add_molecules() to add it?` 
Is there special arrangements needed for registering a second molecule forcefeld?

**Comment by j-wags** [comment]
Thanks for the input files - I'm glad the single-ligand case was validated, and I saw the successful output when I ran locally as well! 

I'd think that openmmforcefields can handle multiple ligands, but I'm not an expert. Are you sure the ligand in `lig2.sdf` and the ligand in your PDB have the same number of Hs? 

Regardless, this is a bit outside our jurisdiction. I'll refer you to the [OpenMMForceFields issue tracker](https://github.com/openmm/openmmforcefields) for the new question, and close this issue since I think the OpenFF part of it is resolved.
